src/transformStateWithClones.js

```diff
@@ -6,8 +6,33 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const history = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        for (const removeElement of action.keysToRemove) {
+          delete stateCopy[removeElement];
+        }
+        break;
+      case 'clear':
+        for (const stateCopyKey in stateCopy) {
+          delete stateCopy[stateCopyKey];
+        }
+        break;
+      default:
+        throw Error('Action is not defined!');
+    }
+    history.push({ ...stateCopy });
```

Comment:
You are pushing a copy of the state to the history array at the end of each loop iteration, which is good. However, to follow the DRY principle, consider creating a new copy of the state at the beginning of each iteration instead of only once before the loop. This way, you can avoid mutating the same object and ensure that each action starts with a fresh state based on the previous one.